File: go/suppressions/suppressions_test.go
package suppressions

import (
	"errors"
	"io"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

const ruleErrcheck = "errcheck"
const ruleFoo = "FOO"
const ruleConstAssign = "CONST_ASSIGN"

func TestParseLine(t *testing.T) {
	tests := []struct {
		name string
		line string
		want []Match
	}{
		{
			name: "plain code no match",
			line: "x := 1",
			want: nil,
		},
		{
			name: "bare nolint",
			line: "x := 1 // nolint",
			want: []Match{{Kind: kindNolint, Rules: nil}},
		},
		{
			name: "nolint with rules",
			line: "x := 1 // nolint: errcheck, gosec",
			want: []Match{{Kind: kindNolint, Rules: []string{ruleErrcheck, "gosec"}}},
		},
		{
			name: "lint ignore with rule",
			line: "const Foo = 1 // lint:ignore CONST_ASSIGN",
			want: []Match{{Kind: kindLintIgnore, Rules: []string{ruleConstAssign}}},
		},
		{
			name: "nolint with whitespace in rules",
			line: "x := 1 // nolint: foo  ,  bar  ",
			want: []Match{{Kind: kindNolint, Rules: []string{"foo", "bar"}}},
		},
		{
			name: "both kinds on one line",
			line: "x := 1 // nolint: errcheck // lint:ignore FOO",
			want: []Match{
				{Kind: kindNolint, Rules: []string{ruleErrcheck}},
				{Kind: kindLintIgnore, Rules: []string{ruleFoo}},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParseLine(tt.line)
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("ParseLine(%q) = %+v, want %+v", tt.line, got, tt.want)
			}
		})
	}
}

func TestScan(t *testing.T) {
	tmp := t.TempDir()

	goFile := filepath.Join(tmp, "a.go")
	if err := os.WriteFile(goFile, []byte(
		"x := 1 // nolint: errcheck\n"+
			"const Foo = 1 // lint:ignore CONST_ASSIGN\n"+
			"y := 2\n",
	), 0o600); err != nil {
		t.Fatal(err)
	}

	txtFile := filepath.Join(tmp, "skip.txt")
	if err := os.WriteFile(txtFile, []byte("// nolint\n"), 0o600); err != nil {
		t.Fatal(err)
	}

	results := Scan(tmp)

	wantNolint := [][]string{{ruleErrcheck}}
	if !reflect.DeepEqual(results[kindNolint], wantNolint) {
		t.Errorf("results[nolint] = %+v, want %+v", results[kindNolint], wantNolint)
	}

	wantLintIgnore := [][]string{{ruleConstAssign}}
	if !reflect.DeepEqual(results[kindLintIgnore], wantLintIgnore) {
		t.Errorf("results[lint_ignore] = %+v, want %+v", results[kindLintIgnore], wantLintIgnore)
	}
}

// captureStdout runs fn and returns everything it wrote to os.Stdout.
func captureStdout(t *testing.T, fn func()) string {
	t.Helper()
	orig := os.Stdout
	r, w, err := os.Pipe()
	if err != nil {
		t.Fatal(err)
	}
	os.Stdout = w
	fn()
	_ = w.Close()
	os.Stdout = orig
	out, err := io.ReadAll(r)
	if err != nil {
		t.Fatal(err)
	}
	return string(out)
}

func TestPrintReport(t *testing.T) {
	t.Run("empty", func(t *testing.T) {
		out := captureStdout(t, func() { PrintReport(map[string][][]string{}) })
		if !strings.Contains(out, "Suppressions: 0 total") {
			t.Errorf("expected zero-total line, got: %q", out)
		}
	})

	t.Run("populated", func(t *testing.T) {
		results := map[string][][]string{
			kindNolint:     {{ruleErrcheck}, {ruleErrcheck, "gosec"}},
			kindLintIgnore: {{ruleFoo}},
		}
		out := captureStdout(t, func() { PrintReport(results) })
		for _, want := range []string{
			"Suppressions: 3 total",
			"nolint: 2",
			"errcheck: 2",
			"lint_ignore: 1",
		} {
			if !strings.Contains(out, want) {
				t.Errorf("expected %q in report, got: %q", want, out)
			}
		}
	})
}

func TestBaselineReadWrite(t *testing.T) {
	tmp := t.TempDir()
	if err := os.WriteFile(filepath.Join(tmp, ".harness-baseline"), []byte(
		"suppressions.nolint 2\ncoverage.min 65\n",
	), 0o600); err != nil {
		t.Fatal(err)
	}

	got, ok := ReadBaseline(tmp)
	if !ok {
		t.Fatal("expected baseline to be read")
	}
	if got["suppressions.nolint"] != 2 || got["coverage.min"] != 65 {
		t.Fatalf("ReadBaseline() = %#v", got)
	}

	results := map[string][][]string{
		kindNolint:     {{ruleErrcheck}},
		kindLintIgnore: {{ruleFoo}},
	}
	if _, err := WriteBaseline(tmp, results, nil); err != nil {
		t.Fatal(err)
	}
	updated, ok := ReadBaseline(tmp)
	if !ok {
		t.Fatal("expected updated baseline")
	}
	if updated["suppressions.nolint"] != 1 ||
		updated["suppressions.lint_ignore"] != 1 ||
		updated["coverage.min"] != 65 {
		t.Fatalf("updated baseline = %#v", updated)
	}
}

func writeBaselineFile(t *testing.T, dir, content string) {
	t.Helper()
	if err := os.WriteFile(filepath.Join(dir, ".harness-baseline"), []byte(content), 0o600); err != nil {
		t.Fatal(err)
	}
}

func measurer(key string, m Measurement) Measurer {
	return Measurer{Key: key, Measure: func() Measurement { return m }}
}

func TestWriteBaselineMergesOverUnknownKeys(t *testing.T) {
	tmp := t.TempDir()
	writeBaselineFile(t, tmp, "custom.thing 7\nmutation.min 40\ncrap.max_violations 9\n")
	measurers := []Measurer{
		measurer("complexity.max_violations", Measured(3)),
		measurer("crap.max_violations", Unavailable("no tests")),
	}

	var written map[string]int
	var err error
	out := captureStdout(t, func() {
		written, err = WriteBaseline(tmp, map[string][][]string{kindNolint: {{ruleErrcheck}}}, measurers)
	})
	if err != nil {
		t.Fatal(err)
	}

	// Unknown keys preserved, measured key written, inapplicable key dropped with a warning.
	if written["custom.thing"] != 7 || written["mutation.min"] != 40 {
		t.Fatalf("unknown keys not preserved: %#v", written)
	}
	if written["complexity.max_violations"] != 3 {
		t.Fatalf("measured key not written: %#v", written)
	}
	if _, has := written["crap.max_violations"]; has {
		t.Fatalf("unavailable key kept: %#v", written)
	}
	if !strings.Contains(out, "crap.max_violations: dropped") {
		t.Fatalf("expected drop warning, got: %q", out)
	}
	want := "complexity.max_violations 3\ncustom.thing 7\nmutation.min 40\nsuppressions.lint_ignore 0\nsuppressions.nolint 1\n"
	if content := formatBaseline(written); content != want {
		t.Fatalf("baseline = %q, want %q", content, want)
	}
}

func TestWriteBaselineZeroesAVanishedSuppressionKind(t *testing.T) {
	tmp := t.TempDir()
	writeBaselineFile(t, tmp, "suppressions.nolint 5\n")

	written, err := WriteBaseline(tmp, map[string][][]string{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if value, has := written["suppressions.nolint"]; !has || value != 0 {
		t.Fatalf("vanished kind not recorded as 0: %#v", written)
	}
}

func TestWriteBaselineRecordsALegitimateZero(t *testing.T) {
	tmp := t.TempDir()

	written, err := WriteBaseline(tmp, nil, []Measurer{measurer("complexity.max_violations", Measured(0))})
	if err != nil {
		t.Fatal(err)
	}
	if value, has := written["complexity.max_violations"]; !has || value != 0 {
		t.Fatalf("measured 0 not recorded: %#v", written)
	}
}

func TestWriteBaselineAbortsAndWritesNothingOnAFailedMeasurement(t *testing.T) {
	tmp := t.TempDir()
	original := "coverage.min 40\n"
	writeBaselineFile(t, tmp, original)
	measurers := []Measurer{
		measurer("complexity.max_violations", Failed("lizard failed to run (exit 2)")),
		measurer("crap.max_violations", Measured(1)),
	}

	written, err := WriteBaseline(tmp, nil, measurers)

	var mErr *MeasurementError
	if !errors.As(err, &mErr) {
		t.Fatalf("expected *MeasurementError, got %v", err)
	}
	if mErr.Key != "complexity.max_violations" || !strings.Contains(mErr.Reason, "exit 2") {
		t.Fatalf("unexpected error: %#v", mErr)
	}
	if written != nil {
		t.Fatalf("expected no baseline on failure, got %#v", written)
	}
	after, _ := ReadBaseline(tmp)
	if len(after) != 1 || after["coverage.min"] != 40 {
		t.Fatalf("baseline file changed on failure: %#v", after)
	}
}

func TestMeasureRatchetedStopsAtTheFirstFailure(t *testing.T) {
	calls := 0
	count := func(m Measurement) func() Measurement {
		return func() Measurement { calls++; return m }
	}
	measured := MeasureRatcheted([]Measurer{
		{Key: "a", Measure: count(Measured(1))},
		{Key: "b", Measure: count(Failed("boom"))},
		{Key: "c", Measure: count(Measured(3))},
	})

	if calls != 2 || len(measured) != 2 || measured[1].Key != "b" || measured[1].Error != "boom" {
		t.Fatalf("calls=%d measured=%#v", calls, measured)
	}
}

func TestBaselineFloor(t *testing.T) {
	tmp := t.TempDir()
	if _, ok := BaselineFloor(tmp, "complexity.max_violations"); ok {
		t.Fatal("expected no floor without a baseline file")
	}
	writeBaselineFile(t, tmp, "coverage.min 0\ncomplexity.max_violations 12\n")
	if _, ok := BaselineFloor(tmp, "crap.max_violations"); ok {
		t.Fatal("expected no floor for an absent key")
	}
	if floor, ok := BaselineFloor(tmp, "complexity.max_violations"); !ok || floor != 12 {
		t.Fatalf("BaselineFloor = %d, %v", floor, ok)
	}
}

func TestCheckBaselineDetectsGrowth(t *testing.T) {
	tmp := t.TempDir()
	if err := os.WriteFile(filepath.Join(tmp, ".harness-baseline"), []byte("coverage.min 0\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	findings := []Finding{{Match: Match{Kind: kindNolint}, Location: "a.go:1"}}

	if CheckBaseline(tmp, findings, true, "go run harness.go suppressions --update-baseline", false) {
		t.Fatal("expected suppression growth to fail")
	}
}

File: bun/tests/suppressions.test.ts
import { afterAll, beforeAll, describe, expect, test } from 'bun:test';
import { mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
  baselineFloor,
  complexityGatesOrWarn,
  coverageMinDefault,
  duplicateBlockCount,
  lizardWarningCount,
  type Measurement,
  measureRatcheted,
  parseLineForSuppressions,
  parseMinArg,
  RATCHETED_METRICS,
  type RatchetedMetric,
  readBaseline,
  runCapture,
  scanSuppressionFindings,
  scanSuppressions,
  writeBaseline,
} from '../harness';

/** A real `lizard -Eduplicate` tail: two blocks, then the footer that proves it ran. */
const DUPLICATE_REPORT = [
  'Duplicates',
  '===================================',
  'Duplicate block:',
  '--------------------------',
  'src/a.ts:24 ~ 28',
  'src/b.ts:50 ~ 55',
  '^^^^^^^^^^^^^^^^^^^^^^^^^^',
  '',
  'Duplicate block:',
  '--------------------------',
  'src/a.ts:24 ~ 28',
  'src/b.ts:60 ~ 66',
  '^^^^^^^^^^^^^^^^^^^^^^^^^^',
  '',
  'Total duplicate rate: 2.97%',
  'Total unique rate: 96.73%',
  '',
].join('\n');

describe('parseLineForSuppressions', () => {
  test('plain code returns no matches', () => {
    expect(parseLineForSuppressions('const x = 1;')).toEqual([]);
  });

  test('@ts-ignore', () => {
    expect(parseLineForSuppressions('// @ts-ignore')).toEqual([{ kind: 'ts-ignore', rules: [] }]);
  });

  test('@ts-expect-error with trailing text', () => {
    expect(parseLineForSuppressions('// @ts-expect-error because reasons')).toEqual([
      { kind: 'ts-expect-error', rules: [] },
    ]);
  });

  test('@ts-nocheck', () => {
    expect(parseLineForSuppressions('// @ts-nocheck')).toEqual([{ kind: 'ts-nocheck', rules: [] }]);
  });

  test('eslint-disable-line with no rules', () => {
    expect(parseLineForSuppressions('foo(); // eslint-disable-line')).toEqual([
      { kind: 'eslint-disable', rules: [] },
    ]);
  });

  test('eslint-disable-next-line with colon-separated rules and whitespace', () => {
    expect(parseLineForSuppressions('// eslint-disable-next-line: no-unused-vars,  semi')).toEqual([
      { kind: 'eslint-disable', rules: ['no-unused-vars', 'semi'] },
    ]);
  });

  test('eslint-disable in block comment', () => {
    expect(parseLineForSuppressions('/* eslint-disable */')).toEqual([
      { kind: 'eslint-disable', rules: [] },
    ]);
  });

  test('biome-ignore with namespaced rule', () => {
    expect(
      parseLineForSuppressions('// biome-ignore lint/style/useSingleVarDeclarator: reason'),
    ).toEqual([{ kind: 'biome-ignore', rules: ['lint/style/useSingleVarDeclarator'] }]);
  });
});

describe('scanSuppressions', () => {
  let tmp: string;

  beforeAll(() => {
    tmp = mkdtempSync(join(tmpdir(), 'bun-suppr-'));
    mkdirSync(join(tmp, 'src'));
    mkdirSync(join(tmp, 'tests'));
    writeFileSync(
      join(tmp, 'src', 'a.ts'),
      '// @ts-ignore\nconst x = 1;\nfoo(); // eslint-disable-line: no-unused-vars\n',
    );
    writeFileSync(
      join(tmp, 'tests', 'b.ts'),
      '// biome-ignore lint/style/useSingleVarDeclarator: reason\nconst y = 2;\n',
    );
    writeFileSync(join(tmp, 'src', 'skip.md'), '// @ts-ignore (should be ignored — not .ts)\n');
  });

  afterAll(() => {
    rmSync(tmp, { recursive: true, force: true });
  });

  test('counts suppressions across src and tests and ignores non-ts files', async () => {
    const results = await scanSuppressions([join(tmp, 'src'), join(tmp, 'tests')]);
    expect(results['ts-ignore']).toHaveLength(1);
    expect(results['eslint-disable']).toEqual([['no-unused-vars']]);
    expect(results['biome-ignore']).toEqual([['lint/style/useSingleVarDeclarator']]);
  });
});

describe('scanSuppressionFindings locations (FIX 5)', () => {
  test('locations are always relative to ROOT, for both scan branches', async () => {
    const tmp = mkdtempSync(join(tmpdir(), 'bun-suppr-loc-'));
    try {
      // A root that is itself a .ts file exercises the single-file branch; a root
      // that is a directory exercises the directory-walk branch. Before the fix,
      // the single-file branch echoed the raw (here: absolute) input path while
      // the directory branch always built an absolute path — a mix of formats.
      // Assembled at runtime, not written as a literal directive comment, so this
      // fixture doesn't trip the real suppression ratchet when this file is itself
      // scanned as part of tests/.
      const ignoreDirective = ['//', '@ts-ignore'].join(' ');
      const topFile = join(tmp, 'top.ts');
      writeFileSync(topFile, `${ignoreDirective}\nconst x = 1;\n`);
      mkdirSync(join(tmp, 'nested'));
      writeFileSync(join(tmp, 'nested', 'child.ts'), `${ignoreDirective}\nconst y = 2;\n`);

      const findings = await scanSuppressionFindings([topFile, join(tmp, 'nested')]);
      expect(findings).toHaveLength(2);
      for (const finding of findings) {
        expect(finding.location.startsWith('/')).toBe(false);
        expect(finding.location).toMatch(/:\d+$/);
      }
    } finally {
      rmSync(tmp, { recursive: true, force: true });
    }
  });
});

describe('baseline helpers', () => {
  const baselineRoots: string[] = [];

  afterAll(() => {
    for (const root of baselineRoots) rmSync(root, { recursive: true, force: true });
  });

  test('readBaseline parses key value lines', async () => {
    const root = mkdtempSync(join(tmpdir(), 'bun-baseline-'));
    baselineRoots.push(root);
    writeFileSync(join(root, '.harness-baseline'), 'suppressions.ts-ignore 2\ncoverage.min 70\n');

    expect(await readBaseline(root)).toEqual({
      'suppressions.ts-ignore': 2,
      'coverage.min': 70,
    });
  });

  test('coverageMinDefault uses flag before baseline', async () => {
    const root = mkdtempSync(join(tmpdir(), 'bun-baseline-'));
    baselineRoots.push(root);
    writeFileSync(join(root, '.harness-baseline'), 'coverage.min 60\n');
    const originalArgv = process.argv;
    try {
      process.argv = ['bun', 'harness.ts', 'coverage'];
      expect(await coverageMinDefault(root)).toBe(60);
      process.argv = ['bun', 'harness.ts', 'coverage', '--min=10'];
      expect(await coverageMinDefault(root)).toBe(10);
    } finally {
      process.argv = originalArgv;
    }
  });
});

describe('writeBaseline merge semantics', () => {
  const roots: string[] = [];

  afterAll(() => {
    for (const root of roots) rmSync(root, { recursive: true, force: true });
  });

  // Every test must pass an explicit base: the default is the template root, and
  // one forgotten argument would rewrite the shipped `.harness-baseline`.
  function project(existing?: string): string {
    const root = mkdtempSync(join(tmpdir(), 'bun-write-baseline-'));
    roots.push(root);
    if (existing !== undefined) writeFileSync(join(root, '.harness-baseline'), existing);
    return root;
  }

  function read(root: string): string {
    return readFileSync(join(root, '.harness-baseline'), 'utf8');
  }

  test('measured floors are written and unknown keys are preserved', async () => {
    const root = project('mutation.min 63\ncoverage.min 40\n');
    const written = await writeBaseline(
      { 'ts-ignore': [[]] },
      { 'coverage.min': { value: 72 }, 'complexity.max_violations': { value: 3 } },
      root,
    );

    expect(written.ok).toBe(true);
    expect(await readBaseline(root)).toMatchObject({
      'mutation.min': 63,
      'coverage.min': 72,
      'complexity.max_violations': 3,
      'suppressions.ts-ignore': 1,
    });
  });

  test('a kind that vanished from the tree ratchets to 0, not to its old count', async () => {
    const root = project('suppressions.ts-ignore 5\nsuppressions.biome-ignore 2\n');
    await writeBaseline({ 'ts-ignore': [[]] }, {}, root);

    expect(await readBaseline(root)).toMatchObject({
      'suppressions.ts-ignore': 1,
      'suppressions.biome-ignore': 0,
    });
  });

  test('an unavailable metric drops its key instead of carrying the shipped number', async () => {
    const root = project('coverage.min 100\n');
    const written = await writeBaseline({}, { 'coverage.min': { unavailable: 'no tests' } }, root);

    expect(written.ok).toBe(true);
    expect(read(root)).not.toContain('coverage.min');
  });

  test('an errored metric writes nothing at all', async () => {
    const root = project('coverage.min 40\n');
    const written = await writeBaseline(
      { 'ts-ignore': [[]] },
      { 'coverage.min': { error: 'the test run under coverage failed (exit 1)' } },
      root,
    );

    expect(written).toEqual({
      ok: false,
      broken: [['coverage.min', 'the test run under coverage failed (exit 1)']],
    });
    expect(read(root)).toBe('coverage.min 40\n');
  });

  test('lines are sorted so the file is diff-stable', async () => {
    const root = project();
    await writeBaseline(
      {},
      { 'crap.max_violations': { value: 2 }, 'coverage.min': { value: 1 } },
      root,
    );

    const keys = read(root)
      .trim()
      .split('\n')
      .map((line) => line.split(' ')[0]);
    expect(keys).toEqual([...keys].sort());
  });
});

describe('measureRatcheted', () => {
  const metric = (key: string, measurement: Measurement): RatchetedMetric => ({
    key,
    measure: () => Promise.resolve(measurement),
  });

  test('stops at the first error so the later noise is not reported', async () => {
    const measured = await measureRatcheted([
      metric('coverage.min', { value: 70 }),
      metric('complexity.max_violations', { error: 'lizard failed to run (exit 2)' }),
      metric('crap.max_violations', { value: 0 }),
    ]);

    expect(Object.keys(measured)).toEqual(['coverage.min', 'complexity.max_violations']);
  });

  test('an unavailable metric does not stop the ones after it', async () => {
    const measured = await measureRatcheted([
      metric('coverage.min', { unavailable: 'no tests' }),
      metric('crap.max_violations', { value: 4 }),
    ]);

    expect(measured['crap.max_violations']).toEqual({ value: 4 });
  });

  test('the shipped table covers exactly the keys this template ratchets', () => {
    expect(RATCHETED_METRICS.map((m) => m.key)).toEqual([
      'coverage.min',
      'complexity.max_violations',
      'duplication.max_blocks',
      'crap.max_violations',
    ]);
  });
});

describe('lizard floors', () => {
  const roots: string[] = [];

  afterAll(() => {
    for (const root of roots) rmSync(root, { recursive: true, force: true });
  });

  function project(existing?: string): string {
    const root = mkdtempSync(join(tmpdir(), 'bun-complexity-'));
    roots.push(root);
    mkdirSync(join(root, 'src'));
    mkdirSync(join(root, 'tests'));
    writeFileSync(join(root, 'src', 'index.ts'), 'export const x = 1;\n');
    writeFileSync(join(root, 'tests', 'x.test.ts'), 'export const y = 1;\n');
    if (existing !== undefined) writeFileSync(join(root, '.harness-baseline'), existing);
    return root;
  }

  test('no floor recorded means report-only, not -i 0', async () => {
    const root = project();
    const [gate] = await complexityGatesOrWarn(root);

    expect(gate?.description).toContain('report-only: no .harness-baseline floor');
    expect(gate?.cmd.at(-1)).toBe('1000000');
    expect(gate?.extract?.('')).toContain('suppressions --update-baseline');
    expect(await baselineFloor('complexity.max_violations', root)).toBeUndefined();
  });

  test('a recorded floor is handed to lizard as -i N', async () => {
    const root = project('complexity.max_violations 4\n');
    const [gate] = await complexityGatesOrWarn(root);

    expect(gate?.description).toBe('Complexity (lizard, baseline 4)');
    expect(gate?.cmd.slice(-2)).toEqual(['-i', '4']);
    expect(await baselineFloor('complexity.max_violations', root)).toBe(4);
  });

  test('a floor of 0 is a real floor, not a missing one', async () => {
    const root = project('complexity.max_violations 0\n');
    const [gate] = await complexityGatesOrWarn(root);

    expect(gate?.description).toBe('Complexity (lizard)');
    expect(gate?.cmd.slice(-2)).toEqual(['-i', '0']);
  });

  // The floor only reproduces against the same target set the complexity gate
  // uses, so compare the two lists instead of hardcoding one of them.
  test('the duplicate-block gate scans the same targets as the complexity gate', async () => {
    const root = project();
    const [complexity, duplication] = await complexityGatesOrWarn(root);
    const targets = (complexity?.cmd ?? []).slice(2, complexity?.cmd.indexOf('-C'));

    expect(targets).toEqual(['src', 'tests']);
    expect(duplication?.cmd.slice(2, 2 + targets.length)).toEqual(targets);
    expect(duplication?.cmd.slice(2 + targets.length)).toEqual([
      '-Eduplicate',
      '-w',
      '-i',
      '1000000',
    ]);
  });

  test('no duplication floor recorded means report-only, never a failing verdict', async () => {
    const root = project();
    const [, gate] = await complexityGatesOrWarn(root);

    expect(gate?.description).toContain('report-only: no .harness-baseline floor');
    const verdict = gate?.verdict?.(DUPLICATE_REPORT);
    expect(verdict?.ok).toBe(true);
    expect(verdict?.detail).toContain('suppressions --update-baseline');
  });

  test('a recorded floor tolerates exactly that many blocks', async () => {
    const root = project('duplication.max_blocks 2\n');
    const [, gate] = await complexityGatesOrWarn(root);

    expect(gate?.description).toBe('Duplicate blocks (lizard, baseline 2)');
    expect(gate?.verdict?.(DUPLICATE_REPORT)).toEqual({ ok: true, detail: '2 block(s)' });
  });

  test('one block over the floor fails', async () => {
    const root = project('duplication.max_blocks 1\n');
    const [, gate] = await complexityGatesOrWarn(root);

    expect(gate?.verdict?.(DUPLICATE_REPORT)?.ok).toBe(false);
  });

  test('a missing report fails the verdict instead of passing at zero', async () => {
    const root = project('duplication.max_blocks 5\n');
    const [, gate] = await complexityGatesOrWarn(root);

    expect(gate?.verdict?.('uvx: command not found')?.ok).toBe(false);
  });

  // Report-only means report-only: a garbled run cannot turn day one red either.
  test('a missing report is still report-only when no floor is recorded', async () => {
    const root = project();
    const [, gate] = await complexityGatesOrWarn(root);

    expect(gate?.verdict?.('uvx: command not found')?.ok).toBe(true);
  });

  // The guard that makes `verdict` safe: a crashed tool prints no findings, and
  // scoring that as "zero findings" is the false-pass the whole design avoids.
  test('a verdict is not consulted when the command exited non-zero', async () => {
    const result = await runCapture({
      description: 'crashes',
      cmd: ['sh', '-c', 'exit 3'],
      verdict: () => ({ ok: true, detail: 'looks clean to me' }),
    });

    expect(result.ok).toBe(false);
    expect(result.exitCode).toBe(3);
  });

  test('a failing verdict on a clean exit still exits non-zero', async () => {
    const result = await runCapture({
      description: 'passes, reports trouble',
      cmd: ['sh', '-c', 'exit 0'],
      verdict: () => ({ ok: false, detail: '3 block(s)' }),
    });

    expect(result.ok).toBe(false);
    expect(result.exitCode).toBe(1);
    expect(result.detail).toBe('3 block(s)');
  });
});

describe('lizardWarningCount', () => {
  const summary = [
    '='.repeat(10),
    'Total nloc   Avg.NLOC  AvgCCN  Avg.token   Fun Cnt  Warning cnt   Fun Rt   nloc Rt',
    '--------------------------------------------------------------------------------',
    '      1200       12.0     3.2      100.0        99            7     0.07      0.12',
  ].join('\n');

  test('reads the Warning cnt column out of the summary row', () => {
    expect(lizardWarningCount(summary)).toBe(7);
  });

  test('no summary row is null, never a silent 0', () => {
    expect(lizardWarningCount('lizard: command not found')).toBeNull();
  });
});

describe('duplicateBlockCount', () => {
  test('counts the block headers, not the file lines under them', () => {
    expect(duplicateBlockCount(DUPLICATE_REPORT)).toBe(2);
  });

  test('a report with no blocks is 0', () => {
    expect(duplicateBlockCount('Duplicates\n===\nTotal duplicate rate: 0.00%\n')).toBe(0);
  });

  test('no report at all is null, never a silent 0', () => {
    expect(duplicateBlockCount('uvx: command not found')).toBeNull();
  });
});

describe('parseMinArg (FIX 4)', () => {
  test('flag absent', () => {
    expect(parseMinArg(['bun', 'harness.ts', 'coverage'])).toEqual({ present: false });
  });

  test('valid integer value', () => {
    expect(parseMinArg(['bun', 'harness.ts', 'coverage', '--min=42'])).toEqual({
      present: true,
      ok: true,
      value: 42,
    });
  });

  test('non-numeric value is reported instead of silently becoming NaN', () => {
    expect(parseMinArg(['bun', 'harness.ts', 'coverage', '--min=abc'])).toEqual({
      present: true,
      ok: false,
      raw: 'abc',
    });
  });

  test('non-integer numeric value is rejected', () => {
    expect(parseMinArg(['bun', 'harness.ts', 'coverage', '--min=50.5'])).toEqual({
      present: true,
      ok: false,
      raw: '50.5',
    });
  });
});
